Make port arg optional in `docker compose port`

crazybolillo opened the pull request:
**What I did**

This comand no longer needs a mandatory port. If no port is provided it will instead list all published ports. This behavior falls in line with the behavior provided by `docker port`.

**Related issue**
Closes #11859.

glours commented on cmd/compose/port.go:
```diff
@@ -40,10 +41,15 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
-		Short: "Print the public port for a port binding",
-		Args:  cobra.MinimumNArgs(2),
+		Use:   "port [OPTIONS] SERVICE [PRIVATE_PORT]",
+		Short: "List port mappings or a specific mapping for the service",
```
```suggestion
		Short: "List port mappings or print the public port for a specific mapping for the service",
```

ndeloof replied:
this is just the same help used in docker/cli

glours commented on cmd/compose/port.go:
```diff
@@ -62,19 +68,45 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
+		if err != nil {
+			return err
+		}
+		printPorts(dockerCli, summaries[opts.index].Publishers, opts)
```
We need to test the size of the `summaries` slice otherwise we would potentially trigger `slice bounds out of range` error

glours commented on cmd/compose/port.go:
```diff
@@ -62,19 +68,45 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
+			return err
+		}
+		printPorts(dockerCli, summaries[opts.index].Publishers, opts)
+		return nil
+	} else {
```
the `else` isn't needed as the following statement is the last one and the `if` section finish with a `return`

glours reviewed (changes requested):
Thanks for your contribution, I have some small suggestions to improve your changes

crazybolillo commented:
Thanks for the review, I have taken care of your observations.

glours reviewed (approved):
LGTM

ndeloof commented on cmd/compose/port.go:
```diff
@@ -40,10 +41,15 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
+		Args:  cobra.RangeArgs(1, 2),
 		PreRunE: Adapt(func(ctx context.Context, args []string) error {
+			if len(args) == 1 {
+				opts.list = true
```
incompatibility with `--index` should be tested

crazybolillo replied:
Should `--index` be prohibited if no specific port was provided?

ndeloof commented on cmd/compose/port.go:
```diff
@@ -62,11 +68,40 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
 	}
+
+	if opts.list {
+		summaries, err := backend.Ps(ctx, projectName, api.PsOptions{Services: []string{service}})
```
better update `backend.Port` so it can manage multiple services

ndeloof commented on cmd/compose/port.go:
```diff
@@ -62,11 +68,40 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
+		if err != nil {
+			return err
+		}
+		if len(summaries) <= opts.index {
```
[]ContainerSummary index is unrelated to replica index in service set by `--index`.

ndeloof commented on cmd/compose/port.go:
```diff
@@ -62,11 +68,40 @@ func portCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service)
+		}
+
+		var format string
+		if strings.Contains(publisher.URL, ":") {
```
would make sense to me `PortPublisher` has a `String()` method to handle this, aligned with docker/cli

ndeloof commented:
while not strictly required, I think we should use this opportunity while revisiting this command to add support for app_protocol and port name (https://github.com/compose-spec/compose-spec/blob/master/05-services.md#long-syntax-3)

crazybolillo commented:
> while not strictly required, I think we should use this opportunity while revisiting this command to add support for app_protocol and port name (https://github.com/compose-spec/compose-spec/blob/master/05-services.md#long-syntax-3)

Wow, did not even know those existed. How do you envision it? Allow filtering? Or just displaying the name and app_protocol if they exist for the given port?

crazybolillo commented:
Went ahead and changed `backend.Port `so it returns a slice of publishers. That should fix other issues like dealing with `--index`. Since I was modifying `PortPublisher` I went ahead and also changed its port members to be uint16 so it falls in line with `types.Port`

I think the only thing left is the question about `--index` incompatibility, and support for `app_protocol` and port name,

ndeloof commented on pkg/compose/port.go:
```diff
@@ -26,18 +26,40 @@ import (
 )
 
-func (s *composeService) Port(ctx context.Context, projectName string, service string, port uint16, options api.PortOptions) (string, int, error) {
+func (s *composeService) Port(ctx context.Context, projectName string, service string, port uint16, options api.PortOptions) (api.PortPublishers, error) {
```
as signature changes, and breaks compatibility, makes sense to _also_ rename `Ports`(plural)

crazybolillo replied:
@ndeloof Sorry to bother but do you have any thoughts on my comments? Should we still perform the renaming?

ndeloof commented:
`--index` is used to select service replica, already implemented in `backend.Port` calling `getSpecifiedContainer`
Adding support for app_protocol would not be that simple, as such metadata only exists in compose model, not in actual resources. Let's keep this for a future improvement
